### PR TITLE
8280041: Retry loop issues in java.io.ClassCache

### PR DESCRIPTION
**Pull request by shipilev**
JDK-8277072 introduced java.io.ClassCache, but there seem to be at least two issues with it:
  - The cache cannot disambiguate between cleared SoftReference and the accidental passing of `null` value; in that case, the retry loop would spin indefinitely;
  - If retry loop would spin several times, every time discovering a cleared SoftReference, it would create and register new SoftReference on the ReferenceQueue. However, it would not *drain* the RQ in the loop; in that case, we might have the unbounded garbage accumulating in RQ; 

Attention @rkennke, @plevart.

Additional testing:
  - [x] Linux x86_64 fastdebug `java/io/ObjectStreamClass`
  - [x] Linux x86_64 fastdebug `tier1`
  - [x] Linux x86_64 fastdebug `tier2`
  - [x] Linux x86_64 fastdebug `tier3`

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8280041](https://bugs.openjdk.java.net/browse/JDK-8280041): Retry loop issues in java.io.ClassCache

### Reviewers
 * [Roman Kennke](https://openjdk.java.net/census#rkennke) (@rkennke - **Reviewer**) ⚠️ Review applies to 49b852de88c7ab1e1776c077be3a26852d18376d
 * [Roger Riggs](https://openjdk.java.net/census#rriggs) (@RogerRiggs - **Reviewer**)
 * [Peter Levart](https://openjdk.java.net/census#plevart) (@plevart - **Reviewer**) ⚠️ Review applies to 0f2dcd0de501ee2f0e667138beaf045226f73c21

### Contributors
 * Peter Levart `<plevart@openjdk.org>`

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/7092/head:pull/7092` \
`$ git checkout pull/7092`

Update a local copy of the PR: \
`$ git checkout pull/7092` \
`$ git pull https://git.openjdk.java.net/jdk pull/7092/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 7092`

View PR using the GUI difftool: \
`$ git pr show -t 7092`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/7092.diff">https://git.openjdk.java.net/jdk/pull/7092.diff</a>

</details>

### CODE REVIEW

**Review by rkennke — APPROVED** [review]
Right, good catches! It seems a bit weird that the constructors of Reference would accept null referents anyway, but it is like it is. *shrugs*
One additional improvement would be to change r.get() == null to r.refersTo(null) to check for cleared reference, that avoids reviving the referent, e.g. in SATB GCs. I leave that up to you.

**Comment by shipilev** [comment]
> One additional improvement would be to change r.get() == null to r.refersTo(null) to check for cleared reference, that avoids reviving the referent, e.g. in SATB GCs. I leave that up to you.

I think that does not work, because we want to strongly reference the `val` for eventual return.

**Comment by rkennke** [comment]
> > One additional improvement would be to change r.get() == null to r.refersTo(null) to check for cleared reference, that avoids reviving the referent, e.g. in SATB GCs. I leave that up to you.
> 
> I think that does not work, because we want to strongly reference the `val` for eventual return.

Ah right. Looks ok then!

**Review by rkennke — APPROVED** [review]

**Review by RogerRiggs — COMMENTED** [review]
Can a test be written to verify the correct handling of nulls.

**Comment by shipilev** [comment]
> Can a test be written to verify the correct handling of nulls.

`java.io.ClassCache` is not public (for a reason), I struggle to find a way to access it from a jtreg test...

**Comment by RogerRiggs** [comment]
It may not be worth it. If not, add the label "noreg-hard".

It is possible for add to the test description the modules to be opened:
```
* @modules java.base/java.io:open
```
jtreg will add the appropriate command line args when it is compiled and run.
There are various examples in existing test/jdk/java/io... tests.

**Comment by shipilev** [comment]
> It may not be worth it. If not, add the label "noreg-hard".

Added.
 
> It is possible for add to the test description the modules to be opened:
> ```
> * @modules java.base/java.io:open
> ```
> 
> jtreg will add the appropriate command line args when it is compiled and run. There are various examples in existing test/jdk/java/io... tests.

It is not about modules protection, it is about the plain Java rule: the class is package-private, you cannot subclass from it, unless you are in the same package, period. In older Java days you would "just" put the test in the same package, but we cannot do it now, because that would be a split package and test code would refuse to load. But there is a way to inject `java.io` classes, apparently! I added two tests in two new commits.

**Review by rkennke — APPROVED** [review]
Still looks good (even better!) to me. Thanks!

**Review by plevart — CHANGES_REQUESTED** [review]
Hi,
Sorry for late join. I'm monitoring core-libs-dev mailing list and didn't see the message there. I stumbled on this PR accidentally...

I just have one comment.

**Comment by plevart** [comment]
So, this patch is fine for making ClassCache more robust as a re-usable component inside JDK (checking for non-null return of computeValue). The 2nd change, that apparently shields against unbounded accumulation of garbage, might not be enough in the sense that when garbage accumulated without bounds, there would be not progress. So before the patch, we would at least get an OOME and the loop would end. Now the loop could spin indefinitely.
So perhaps we must rather prevent this from happening in some other way...
I have an idea. Stay tuned.

**Comment by shipilev** [comment]
> So, this patch is fine for making ClassCache more robust as a re-usable component inside JDK (checking for non-null return of computeValue). The 2nd change, that apparently shields against unbounded accumulation of garbage, might not be enough in the sense that when garbage accumulated without bounds, there would be not progress. So before the patch, we would at least get an OOME and the loop would end. Now the loop could spin indefinitely. So perhaps we must rather prevent this from happening in some other way... I have an idea. Stay tuned.

I think playing the catch-up race with GC (which we will eventually win, IMO) is better than obscure OOME. Note this matches other places we do the weak-reference loops, for example in `MethodType`: https://github.com/openjdk/jdk/blob/master/src/java.base/share/classes/java/lang/invoke/MethodType.java#L1390-L1401

**Comment by plevart** [comment]
> > So, this patch is fine for making ClassCache more robust as a re-usable component inside JDK (checking for non-null return of computeValue). The 2nd change, that apparently shields against unbounded accumulation of garbage, might not be enough in the sense that when garbage accumulated without bounds, there would be not progress. So before the patch, we would at least get an OOME and the loop would end. Now the loop could spin indefinitely. So perhaps we must rather prevent this from happening in some other way... I have an idea. Stay tuned.
> 
> I think playing the catch-up race with GC (which we will eventually win, IMO) is better than obscure OOME. Note this matches other places we do the weak-reference loops, for example in `MethodType`: https://github.com/openjdk/jdk/blob/master/src/java.base/share/classes/java/lang/invoke/MethodType.java#L1390-L1401

I do think that processing enqueued references more promptly is a correct move, so I don't oppose this patch. I'm merely saying that the concern about indefinite growing of reference queue is the same as concern about indefinite looping. In normal circumstances neither would be a problem. But if we are considering special circumstances, we could as well fix both potential problems.

For example, if construction of `CacheRef` object that stays referenced by ClassCache is guaranteed to be followed by at least one `.get()` invocation, the following variant of CacheRef would guarantee progress:

```
    private static class CacheRef<T> extends SoftReference<T> {
        private Class<?> type;
        private T referent;

        CacheRef(T referent, ReferenceQueue<T> queue, Class<?> type) {
            super(referent, queue);
            this.referent = referent;
            this.type = type;
        }

        @Override
        public T get() {
            T referent = this.referent;
            if (referent == null) {
                return super.get();
            } else {
                this.referent = null;
                return referent;
            }
        }
    }
```

WDYT?

**Comment by plevart** [comment]
> Note this matches other places we do the weak-reference loops, for example in `MethodType`: https://github.com/openjdk/jdk/blob/master/src/java.base/share/classes/java/lang/invoke/MethodType.java#L1390-L1401

This one does not have the same problem. It does guarantee progress. But it has another problem. This variant busy-loops while waiting for cleared WeakRefefence to be eventually enqueued by GC / Reference enqueue-ing thread. It could, in addition, pro-actively remove the entry from the map as soon as it finds the WeakReference cleared (with Map.remove(key, value)).

**Comment by shipilev** [comment]
> WDYT?

I like the idea of holding to a value strongly for a brief period of time, in order to guarantee progress! The sample code was a bit hard to follow, so I rewrote the loop a bit with comments, see new commit. This still passes tests.

**Review by plevart — APPROVED** [review]
This looks good, although I don't know whether the additional check for strongReferent != null is needed in clearStrong(). This is all racy code and you have already got a non-null return from getStrong() in case you are calling clearStrong()...

**Comment by shipilev** [comment]
> This looks good, although I don't know whether the additional check for strongReferent != null is needed in clearStrong(). This is all racy code and you have already got a non-null return from getStrong() in case you are calling clearStrong()...

This seems like a standard thing to do to avoid multi-threaded write contention under the race. By the time one thread calls `clearStrong`, some other thread might have already cleared it, and we don't have to clear again. Granted, that a micro-optimization, but I'd like to avoid surprises on this path :)

**Comment by shipilev** [comment]
@RogerRiggs, are you happy with tests?

**Comment by shipilev** [comment]
> > This looks good, although I don't know whether the additional check for strongReferent != null is needed in clearStrong(). This is all racy code and you have already got a non-null return from getStrong() in case you are calling clearStrong()...
> 
> This seems like a standard thing to do to avoid multi-threaded write contention under the race. By the time one thread calls `clearStrong`, some other thread might have already cleared it, and we don't have to clear again. Granted, that is a micro-optimization, but I'd like to avoid surprises on this path :)

Actually, the window after the first `strongVal != null` check is not that large that a second null check would matter. Dropped the null check in `clearStrong` for simplicity. (IIRC, it mattered a bit in prior internal iterations of my patch).

**Review by RogerRiggs — APPROVED** [review]
Looks good, thanks for the tests.

**Comment by shipilev** [comment]
All right, we seem to be good to go.

/integrate
